docs/proposals/build-hooks.md

```diff
@@ -0,0 +1,218 @@
+
+4. Post-build
+
+
+## Alternatives
+
+Existing functionality that could be used to address the use cases above:
+
+- Kubernetes' container lifecycle hooks
+- `assemble` script (for S2I buils)
+- Changes in Dockerfile (for Docker build)
+- Deployment hooks
+- Events API and watchers
+- Custom build
+
+New functionality:
+
+- Build hooks
+
+
+## Discussion on alternatives
+
+Kubernetes' pre-stop hook is not a solution for any of the use cases, because
+it is only [executed prior to a SIGTERM when Kubernetes is about to terminate
+a container](https://github.com/GoogleCloudPlatform/kubernetes/blob/596a8a40d12498b5335140f50753980bfaea4f6b/docs/user-guide/production-pods.md#lifecycle-hooks-and-termination-notice).
+The builder container terminates itself, thus any pre-stop hook would
+be ignored.
+
+### 1. Build pipeline
+
+For *S2I* builds, one might think that providing a custom `assemble`
+script would allow for custom post build actions. However, that's not the case,
+because at the time the script runs, the final state (success or failure) of the
+build is unknown.
+
+If the application defines a DeploymentConfig with an ImageChange trigger, one
+can use the post-deployment hook to trigger a build. This is particularly
+inelegant, fragile and hard to debug when things go wrong:
+
+```json
+{
+  "kind": "DeploymentConfig",
+  "apiVersion": "v1",
+  "metadata": {
+    "name": "chain-1-frontend"
+  },
+  "spec": {
+    "strategy": {
+      "type": "Rolling",
+      "rollingParams": {
+        "pre": {
+          "failurePolicy": "Abort",
+          "execNewPod": {
+            "containerName": "chained-builds",
+            "command": [
+              "sh", "-c", "WEBHOOK_URL=\"https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT/osapi/v1beta3/namespaces/$TARGET_BUILDCONFIG_NAMESPACE/buildconfigs/$TARGET_BUILDCONFIG_NAME/webhooks/$GENERIC_WEBHOOK_SECRET/generic\"; test `curl -skXPOST -w %{http_code} $WEBHOOK_URL` = 200 || >&2 echo -e \"FAILURE: call to generic webhook trigger for BuildConfig $TARGET_BUILDCONFIG_NAMESPACE/$TARGET_BUILDCONFIG_NAME failed. Webhook URL: $WEBHOOK_URL\""
+            ],
+            "env": [
+              {
+                "name": "TARGET_BUILDCONFIG_NAMESPACE",
+                "value": "demo"
+              },
+              {
+                "name": "TARGET_BUILDCONFIG_NAME",
+                "value": "chained-builds-app42"
+              },
+              {
+                "name": "GENERIC_WEBHOOK_SECRET",
+                "value": "${GENERIC_WEBHOOK_SECRET}"
+              }
+            ]
+          }
+        }
+      }
+    },
+    "..."
+  }
+}
+```
+
+Obviously, the other downside is that it involves having a deploy just for the
+sake of triggering another build.
+
+For *Docker* builds, changing the Dockerfile is similar to changing the
+`assemble` script in *S2I* builds, and has the same problem that the final build
+state is unknown at the time instructions in the Dockerfile are executed.
```

Comment:
i'd say changing the dockerfile is more restrictive..you can't really run arbitrary actions within the dockerfile, reasonably (eg calling a webhook).  technically i suppose you could do a RUN for that, but that seems pretty weird and i'm sure it would run into problems.